bestie/server/bigquery_metrics.go

```diff
@@ -0,0 +1,288 @@
+	dat["invocation_id"] = stream.invocationId
+	dat["invocation_sha"] = stream.invocationSha
+	dat["run"] = stream.run
+	dat["test_name"] = stream.testName
+	tags, err := json.Marshal(dat)
+	if err != nil {
+		return bigQueryMetric{}, fmt.Errorf("Error converting JSON to string: %s", err)
+	}
+
+	// Change the metric timestamp from integer to formatted string,
+	// since that is what BigQuery expects for a TIMESTAMP column.
+	//
+	// The incoming timestamp is UTC nanoseconds since epoch, but
+	// storing with microsecond granularity in BigQuery.
+	ts := m.timestamp / 1000
+	tsf := time.Unix(ts/1000000, ts%1000000).Format("2006-01-02 15:04:05.000000")
+
+	return bigQueryMetric{
+		metricname: m.metricname,
+		tags:       string(tags),
+		value:      m.value,
+		timestamp:  tsf,
+	}, nil
+}
+
+// Upload this set of metrics to the specified BigQuery table.
+func uploadTestMetrics(w io.Writer, stream bazelStream, r *metricTestResult) error {
+	// Normalize the BigQuery table identifier based on whether one was specified
+	// in the protobuf message.
+	normalizeTableRef(&r.table)
+	fmt.Fprintf(w, "Normalized table ref: %s\n", r.table.formatTableId())
+
+	// Get client context for this BigQuery operation.
+	ctx := context.Background()
+	client, err := bq.NewClient(ctx, r.table.project)
+	if err != nil {
+		return fmt.Errorf("bigquery.NewClient: %v", err)
+	}
+	defer client.Close()
+	session := bigQuerySession{ctx: ctx, client: client}
+
+	// Check if the BigQuery dataset and table exists; create each as needed.
+	if exist := r.table.isTableExist(session); !exist {
+		ServiceStats.incrementBigQueryTableNotFound()
+		if err := createTable(w, r.table, session); err != nil {
+			return err
+		}
+	}
+
+	// Prepare the metric rows for uploading to BigQuery.
+	// Make sure each row element references its own array item.
+	// Note: bigQueryMetric implements the ValueSaver interface.
+	var bqMetrics []bigQueryMetric
+	var rows []*bigQueryMetric
+	idx := 0
+	for _, m := range r.metrics {
+		// TODO: Decide whether to override client metric time with BES server current time.
+		bqMetric, err := translateMetric(stream, m)
+		if err != nil {
+			ServiceStats.incrementBigQueryMetricDiscard(1)
+			fmt.Fprintf(w, "Discarding metric %s due to error: %s\n", m, err)
+			continue
+		}
+		bqMetrics = append(bqMetrics, bqMetric)
+		rows = append(rows, &bqMetrics[idx])
+		idx++
+	}
+	fmt.Fprintf(w, "\nTranslated metrics for BigQuery upload:\n")
+	for _, bqMetric := range bqMetrics {
+		fmt.Fprintf(w, "  %v\n", bqMetric)
+	}
+	fmt.Fprintln(w)
+
+	// Attempt to upload the metrics, assuming the dataset and table
+	// both exist. If a "not found" error occurs, sleep for a while
+	// then try uploading again.
+	//
+	// Using a finite delay loop here to give BigQuery time to instantiate
+	// the table, which can take a relativly long time (i.e. tens of seconds).
+	ok := false
+	insertionDelay := 0
+	sleepTime := 10
```

Comment:
I don't think having lots of code to try-create tables and datasets is worth it, especially if it includes arbitrary sleeps. Can we make the simplifying assumption that the tables and datasets will exist before we start bestie and point it at them, or else bestie is allowed to error? This simplifies the code greatly, and I think is fine since we're going to make a constant number of tables/datasets (new tables is not dynamic)

Reply:
I agree that this would simplify the endpoint quite a bit. I do want to retain the ability for overriding the default BigQuery dataset and/or tablename that is used, but with the caveat that these must be manually created in BigQuery beforehand.